common/src/main/java/io/pravega/common/util/OrderedItemProcessor.java

```diff
@@ -0,0 +1,253 @@
+    private final Object stateLock = new Object();
+
+    /**
+     * Guards access to the individual item's processors. No two invocations of the processor may run at the same time.
+     */
+    private final Object processingLock = new Object();
+    @GuardedBy("stateLock")
+    private int activeCount;
+    @GuardedBy("stateLock")
+    private boolean closed;
+    @GuardedBy("stateLock")
+    private ReusableLatch emptyNotifier;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the OrderedItemProcessor class.
+     *
+     * @param capacity  The maximum number of concurrent executions.
+     * @param processor A Function that, given an Item, returns a CompletableFuture that indicates when the item has been
+     *                  processed (successfully or not).
+     * @param executor  An Executor for async invocations.
+     */
+    public OrderedItemProcessor(int capacity, Function<ItemType, CompletableFuture<ResultType>> processor, Executor executor) {
+        Preconditions.checkArgument(capacity > 0, "capacity must be a non-negative number.");
+        this.capacity = capacity;
+        this.processor = Preconditions.checkNotNull(processor, "processor");
+        this.executor = Preconditions.checkNotNull(executor, "executor");
+        this.pendingItems = new ArrayDeque<>();
+        this.activeCount = 0;
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    @SneakyThrows(Exception.class)
+    public void close() {
+        ReusableLatch waitSignal = null;
+        synchronized (this.stateLock) {
+            if (this.closed) {
+                return;
+            }
+
+            this.closed = true;
+            if (this.activeCount != 0 || !this.pendingItems.isEmpty()) {
+                // Setup a latch that will be released when the last item completes.
+                this.emptyNotifier = new ReusableLatch(false);
+                waitSignal = this.emptyNotifier;
+            }
+        }
+
+        if (waitSignal != null) {
+            // We have unfinished items. Wait for them.
+            waitSignal.await(CLOSE_TIMEOUT_MILLIS);
+        }
+    }
+
+    //endregion
+
+    //region Processing
+
+    /**
+     * Processes the given item.
+     * * If the OrderedItemProcessor is below capacity, then the item will be processed immediately and the returned
+     * future is the actual result from the Processor.
+     * * If the max capacity is exceeded, the item will be queued up and it will be processed when capacity allows, after
+     * all the items added before it have been processed.
+     * * If an item before this one failed to process and this item is still in the queue (it has not yet been processed),
+     * the returned future will be cancelled with an IllegalStateException to prevent out-of-order executions.
+     * * This method guarantees ordered execution as long as its invocations are serial. That is, it only guarantees that
```

Comment:
This mimics what BookKeeper LedgerHandle does and should be there regardless of the actual implementation details (this was a requirement as to how this class should function). 

I wanted it to have the same behavior. ProcessingLock is there just to ensure that executions do actually occur in order and that JVM thread scheduling (and bad luck) won't cause an out-of-order execution.